### PR TITLE
Fix Issues with Type Lookup Service

### PR DESCRIPTION
**Pull request by iguessthislldo**
Building on #2130, fixes issues with Type Lookup Service. Main improvement is handling the case when the Type Lookup Service writers are invoked when the user entities associate, but the type lookup service entities haven't finished associating. It does this by deferring sending samples until the association is complete. Other improvements include logging and fixing some typos.

The other major changes in this PR are convenience functions for creating GUIDs like `make_part_guid` and convenience functions for converting sequence numbers: ` DCPS::SequenceNumber to_opendds_seqnum(const RTPS::SequenceNumber_t&)`
and `RTPS::SequenceNumber_t to_rtps_seqnum(const DCPS::SequenceNumber&)`.

### CODE REVIEW

File: dds/DCPS/RTPS/Sedp.cpp [hunk 1]
```diff
@@ -3450,6 +3474,53 @@ Sedp::Endpoint::get_builtin_subscriber() const
...
+    break;
+
+  case DCPS::ENTITYKIND_BUILTIN_READER_WITH_KEY:
+    rv.entityKind = DCPS::ENTITYKIND_BUILTIN_WRITER_WITH_KEY;
+    break;
+
+  default:
+    if (DCPS::DCPS_debug_level) {
+      ACE_ERROR((LM_ERROR, "(%P|%t) ERROR: Sedp::Endpoint::counterpart_entity_id: "
+        "Unexpected value: %u\n", rv));
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp:3500]: This is failing in safety profile but may have problems on all configurations - `rv` is not of a type that can be passed to `ACE_Log_Msg` varargs logging.

File: dds/DCPS/DataReaderImpl.cpp [hunk 2]
```diff
@@ -3351,14 +3340,21 @@ DDS::ReturnCode_t DataReaderImpl::setup_deserialization()
...
-    ACE_DEBUG((LM_DEBUG, ACE_TEXT("(%P|%t) ")
-               ACE_TEXT("DataReaderImpl::setup_deserialization: ")
-               ACE_TEXT("Setup successfully with data representations: ")));
-    OPENDDS_SET(Encoding::Kind)::iterator it = decoding_modes_.begin();
+    OPENDDS_STRING encodings;
+    EncodingKinds::iterator it = decoding_modes_.begin();
+    bool first = true;
     for (; it != decoding_modes_.end(); ++it) {
-      ACE_DEBUG((LM_DEBUG, ACE_TEXT("%C "), Encoding::kind_to_string(*it).c_str()));
+      if (first) {
```
Comment (mitza-oci) [inline dds/DCPS/DataReaderImpl.cpp:3347]: If you use `string.empty()` tracking this state separately is not needed

File: dds/DCPS/GuidUtils.h [hunk 3]
```diff
@@ -208,6 +208,36 @@ inline DCPS::RepoId make_id(const DCPS::RepoId& participant_id, const EntityId_t
   return id;
 }
 
+OpenDDS_Dcps_Export inline GUID_t make_guid(
+  const GUID_t& guid, const EntityId_t& entity)
```
Comment (mitza-oci) [inline dds/DCPS/GuidUtils.h:212]:
```suggestion
OpenDDS_Dcps_Export inline
GUID_t make_guid(const GUID_t& guid, const EntityId_t& entity)
```

File: dds/DCPS/RTPS/ParameterListConverter.cpp [hunk 4]
```diff
@@ -462,27 +462,18 @@ bool to_param_list(const ParticipantProxy_t& proxy,
...
   gp_param._d(PID_PARTICIPANT_GUID);
   add_param(param_list, gp_param);
 
   Parameter vid_param;
   vid_param.vendor(proxy.vendorId);
   add_param(param_list, vid_param);
 
-  if (proxy.expectsInlineQos != false)
-  {
+  if (!proxy.expectsInlineQos) {
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/ParameterListConverter.cpp:474]:
```suggestion
  if (proxy.expectsInlineQos) {
```

File: dds/DCPS/RTPS/Sedp.cpp [hunk 5]
```diff
@@ -257,10 +257,14 @@ using DCPS::TimeDuration;
...
 using DCPS::Serializer;
 using DCPS::Encoding;
-using DCPS::ENDIAN_BIG;
-using DCPS::ENDIAN_LITTLE;
 
-const Encoding sedp_encoding(Encoding::KIND_XCDR1, ENDIAN_LITTLE);
+namespace {
+  const Encoding sedp_encoding(Encoding::KIND_XCDR1, DCPS::ENDIAN_LITTLE);
+#ifdef OPENDDS_SECURITY
+  const Encoding secbinprop_encoding(Encoding::KIND_XCDR1, DCPS::ENDIAN_BIG);
```
Comment (mitza-oci) [inline dds/DCPS/RTPS/Sedp.cpp:264]: It's unclear from the name what this is

**Review by mitza-oci — CHANGES_REQUESTED** [review]

File: dds/DCPS/transport/framework/TransportClient.cpp [hunk 6]
```diff
@@ -1092,12 +1092,22 @@ TransportClient::remove_all_msgs()
...
 
-void
-TransportClient::terminate_send_if_suspended() {
+void TransportClient::terminate_send_if_suspended()
+{
   links_.terminate_send_if_suspended();
 }
 
+bool TransportClient::associated_with(const GUID_t& remote) const
+{
```
Comment (jrw972) [inline dds/DCPS/transport/framework/TransportClient.cpp:1101]: Locking.

File: dds/DCPS/transport/framework/TransportClient.cpp [hunk 7]
```diff
@@ -1092,12 +1092,22 @@ TransportClient::remove_all_msgs()
...
+  return data_link_index_.count(remote);
 }
+
+bool TransportClient::pending_association_with(const GUID_t& remote) const
+{
+  return pending_.count(remote);
```
Comment (jrw972) [inline dds/DCPS/transport/framework/TransportClient.cpp:1107]: Locking.

**Review by jrw972 — CHANGES_REQUESTED** [review]

**Review by kuznetsovmoci — CHANGES_REQUESTED** [review]

File: dds/DCPS/RTPS/Sedp.cpp (hunk 5 above)
Comment (iguessthislldo) [reply dds/DCPS/RTPS/Sedp.cpp:264]: Yeah, I thought `security_binary_property_encoding` was thought was a little long, but looking at it now I guess it's not too bad.

File: dds/DCPS/RTPS/MessageTypes.h [hunk 8]
```diff
@@ -29,14 +29,12 @@ namespace OpenDDS {
...
-        { 0x00, 0x00, 0x00, 0x00 },
-      PARTICIPANT_MESSAGE_DATA_KIND_AUTOMATIC_LIVELINESS_UPDATE =
-        { 0x00, 0x00, 0x00, 0x01 },
-      PARTICIPANT_MESSAGE_DATA_KIND_MANUAL_LIVELINESS_UPDATE =
-        { 0x00, 0x00, 0x00, 0x02 };
-
+    const DCPS::EntityId_t PARTICIPANT_MESSAGE_DATA_KIND_UNKNOWN =
+      {{0x00, 0x00, 0x00}, 0x00};
+    const DCPS::EntityId_t PARTICIPANT_MESSAGE_DATA_KIND_AUTOMATIC_LIVELINESS_UPDATE =
+      {{0x00, 0x00, 0x00}, 0x00};
```
[the remark this answers is not in the document]
Comment (iguessthislldo) [reply dds/DCPS/RTPS/MessageTypes.h:35]: Good catch

File: dds/DCPS/RTPS/Sedp.cpp (hunk 5 above)
Comment (iguessthislldo) [reply dds/DCPS/RTPS/Sedp.cpp:264]: Actually noting is using this... I guess I'll remove it?

File: dds/DCPS/GuidUtils.h [hunk 9]
```diff
@@ -208,6 +208,36 @@ inline DCPS::RepoId make_id(const DCPS::RepoId& participant_id, const EntityId_t
...
+}
+
+OpenDDS_Dcps_Export inline
+GUID_t make_part_guid(const GUID_t& guid)
+{
+  return make_part_guid(guid.guidPrefix);
+}
+
+OpenDDS_Dcps_Export inline
+GUID_t make_unknown_guid(const GuidPrefix_t& prefix)
```
[the remark this answers is not in the document]
Comment (iguessthislldo) [reply dds/DCPS/GuidUtils.h:230]: How about `make_*_entity_guid`?

**Review by mitza-oci — APPROVED** [review]

**Review by kuznetsovmoci — APPROVED** [review]

**Pull request merged**